controller/api/destination/dst_overrides_adaptor.go

```diff
@@ -0,0 +1,51 @@
+package destination
+
+import (
+	"fmt"
+
+	"github.com/linkerd/linkerd2/controller/api/destination/watcher"
+	sp "github.com/linkerd/linkerd2/controller/gen/apis/serviceprofile/v1alpha2"
+	"k8s.io/apimachinery/pkg/api/resource"
+)
+
+// dstOverridesAdaptor holds an underlying ProfileUpdateListener and updates
+// that listener with changes to the dstOverrides field of serviceProfiles.
+type dstOverridesAdaptor struct {
```

Comment:
rather than build a brand new adapter, I think this logic should be built into profileTranslator directly.  In particular, the profileTranslator already holds the `fullyQualifiedName` which I believe should be used as the default DstOverride if the list of DstOverrides is empty.

Reply:
@adleong Yep, To remove this adaptor, The following logic would be part of the translator's `toServiceProfile` method
```go
	// If there are no destination overrides set and fqn is known,
	// return a destination override  so that it's known the host is a service.
	if len(profile.Spec.DstOverrides) == 0 && pt.fullyQualifiedName != "" {
		dst := &sp.WeightedDst{
			Authority: fmt.Sprintf("%s.:%d", pt.fullyQualifiedName, pt.port),
			Weight:    resource.MustParse("1"),
		}
		profile.Spec.DstOverrides = []*sp.WeightedDst{dst}
	}
```

Note that `pt.fullyQualifiedName != ""` is needed as the same translator is also used when the IP does not resolve to a service  FQDN i.e at https://github.com/linkerd/linkerd2/blob/tarun/remove-ts/controller/api/destination/server.go#L246-L247 where we do not add the `dstOverrides`

By doing this, It felt like we were leaking higher lever logic into the lower level translator abstraction and hence having adaptor feels much better as in the other case, this adaptor is just not attached with the translator without having to have this logic in the translator. WDYT?

Reply:
This doesn't seem like leaky logic to me.  I think the profile translator is the correct and appropriate place for this logic.